### PR TITLE
Visitor can Register for an account

### PR DESCRIPTION
**Pull request by HannaSwDn**
- Adds Devise gem
- Adds feature file for registration
- Creates User model
- Creates User controller

## Link to PT:
https://www.pivotaltracker.com/story/show/163088269

## Registration view:
screen shot 2019-01-11 at 10 28 50

### CODE REVIEW

File: app/controllers/user_controller.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class UserController < ApplicationController
```
Comment (GergKllai1) [inline app/controllers/user_controller.rb]: Can you try removing this? I don't think that you need the UserController to make this work

File: app/views/users/confirmations/new.html.erb [hunk 2]
```diff
@@ -0,0 +1,16 @@
+<h2>Resend confirmation instructions</h2>
```
Comment (GergKllai1) [inline app/views/users/confirmations/new.html.erb]: You don't need to generate all the views, devise comes bundled with them you can use them without this folder.
You only need to generate the one that you want to change!
Delete this please

File: app/controllers/user_controller.rb (hunk 1 above)
Comment (HannaSwDn) [reply app/controllers/user_controller.rb]: Done

File: app/views/users/confirmations/new.html.erb (hunk 2 above)
Comment (HannaSwDn) [reply app/views/users/confirmations/new.html.erb]: Done

File: app/views/users/sessions/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,26 @@
+<h2>Log in</h2>
```
Comment (oliverochman) [inline app/views/users/sessions/new.html.erb]: why do you need this one?

File: app/views/users/registrations/edit.html.erb [hunk 4]
```diff
@@ -0,0 +1,43 @@
+<h2>Edit <%= resource_name.to_s.humanize %></h2>
```
Comment (GergKllai1) [inline app/views/users/registrations/edit.html.erb]: For now we don't need edit, please remove it

Comment (oliverochman) [inline app/views/users/registrations/edit.html.erb]: why do you need this one?

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+class User < ApplicationRecord
+  
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable,
+         :omniauthable
```
Comment (oliverochman) [inline app/models/user.rb]: do you use omniauth?

File: app/views/users/sessions/new.html.erb (hunk 3 above)
Comment (GergKllai1) [inline app/views/users/sessions/new.html.erb]: Have you changed anything here from the default? If not please remove this.

File: app/views/users/shared/_links.html.erb [hunk 6]
```diff
@@ -0,0 +1,25 @@
+<%- if controller_name != 'sessions' %>
```
Comment (oliverochman) [inline app/views/users/shared/_links.html.erb]: why do you need this one?

File: config/initializers/devise.rb [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+  # change their passwords.
+  config.reset_password_within = 6.hours
+
+  # When set to false, does not sign a user in automatically after their password is
+  # reset. Defaults to true, so a user is signed in automatically after a reset.
+  # config.sign_in_after_reset_password = true
+
+  # ==> Configuration for :encryptable
+  # Allow you to use another hashing or encryption algorithm besides bcrypt (default).
+  # You can use :sha1, :sha512 or algorithms from others authentication tools as
```
Comment (oliverochman) [inline config/initializers/devise.rb]: clean this file up

File: spec/models/user_spec.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
```
Comment (oliverochman) [inline spec/models/user_spec.rb]: add unit tests for the user

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: features/step_definitions/visitor_can_register_for_an_account.feature [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    Background:
+        Given the following user exists
+        | name | email          | password | password_confirmation |
+        | Bill | bill@email.com | password | password              |
+        Given I visit the landing page
+        And I click 'Register'
+
+    Scenario: When I fill in the sign up form correctly, I can sign up
+        Given I fill in 'Email' field with 'mail@email.com'
+        And I fill in 'Password' field with 'password'
```
Comment (oliverochman) [inline features/step_definitions/visitor_can_register_for_an_account.feature]: you have added a name field to the registration, why dont you test that one here?

File: features/step_definitions/visitor_can_register_for_an_account.feature [hunk 10]
```diff
@@ -0,0 +1,19 @@
+Feature: Visitor can register for an account
+
+    As a Visitor,
+    In order to access unique features on the news app,
+    I would like to be able to register for an account on the site.
+
+    Background:
+        Given the following user exists
+        | name | email          | password | password_confirmation |
+        | Bill | bill@email.com | password | password              |
```
Comment (oliverochman) [inline features/step_definitions/visitor_can_register_for_an_account.feature]: Is there any need for this background step

File: features/step_definitions/visitor_can_register_for_an_account.feature [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+        Given I visit the landing page
+        And I click 'Register'
+
+    Scenario: When I fill in the sign up form correctly, I can sign up
+        Given I fill in 'Email' field with 'mail@email.com'
+        And I fill in 'Password' field with 'password'
+        And I fill in 'Password confirmation' field with 'password'
+        And I click 'Sign up'
+        Then I should see 'Welcome! You have signed up successfully.'
```
Comment (oliverochman) [inline features/step_definitions/visitor_can_register_for_an_account.feature]: You need to test some sad paths

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: app/views/users/sessions/new.html.erb (hunk 3 above)
Comment (HannaSwDn) [reply app/views/users/sessions/new.html.erb]: We do not, removed

File: app/views/users/registrations/edit.html.erb (hunk 4 above)
Comment (HannaSwDn) [reply app/views/users/registrations/edit.html.erb]: We do not, removed

File: app/models/user.rb (hunk 5 above)
Comment (HannaSwDn) [reply app/models/user.rb]: Not right now, we will remove it

File: app/views/users/shared/_links.html.erb (hunk 6 above)
Comment (HannaSwDn) [reply app/views/users/shared/_links.html.erb]: We do not, removed

File: config/initializers/devise.rb (hunk 7 above)
Comment (HannaSwDn) [reply config/initializers/devise.rb]: Done

File: features/step_definitions/visitor_can_register_for_an_account.feature [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+    Scenario: When password is too short, I get an error message
```
Comment (GergKllai1) [inline features/step_definitions/visitor_can_register_for_an_account.feature]: You can have in every sad path in one scenario, similar to this example:
```
Scenario: When I don't fill in the sign up form I get error messages
        When I click 'Sign up'
        Then I should see "Email can't be blank"
        And I should see "Password can't be blank"
        And I should see "Name can't be blank"
        And I should see "Address can't be blank"
        And I should see "Phone can't be blank"

 Scenario: When Password and Password confirmation doesn't match I get error message
        When I fill in 'Password' field with 'password'
        And I fill in 'Password confirmation' field with 'pasword'
        And I click 'Sign up'
        Then I should see "Password confirmation doesn't match Password"

    Scenario: When Email is taken I get error message
        When I fill in 'Email' field with 'real@email.com'
        And I click 'Sign up'
        Then I should see "Email has already been taken"
```

File: app/views/partials/_navbar.html.haml [hunk 13]
```diff
@@ -1,3 +1,6 @@
+= notice || alert
 =link_to "The Sunrise Press", root_path
-=link_to "Registration"
-=link_to "Login"
+=link_to "Register", new_user_registration_path
```
Comment (GergKllai1) [inline app/views/partials/_navbar.html.haml]: Can you put this into the `else` part of the `if` statement so that it only appears if you are not logged in?

File: app/views/partials/_navbar.html.haml [hunk 14]
```diff
@@ -1,3 +1,6 @@
...
+-if user_signed_in?
+    =link_to "Logout", destroy_user_session_path, method: :delete
+=link_to "Login"
```
Comment (GergKllai1) [inline app/views/partials/_navbar.html.haml]: Same as above

**Review by GergKllai1 — CHANGES_REQUESTED** [review]

File: features/step_definitions/visitor_can_register_for_an_account.feature (hunk 10 above)
Comment (HannaSwDn) [reply features/step_definitions/visitor_can_register_for_an_account.feature]: Now there is

File: features/step_definitions/visitor_can_register_for_an_account.feature (hunk 9 above)
Comment (HannaSwDn) [reply features/step_definitions/visitor_can_register_for_an_account.feature]: Name field is removed since we do not need it at this point

File: features/step_definitions/visitor_can_register_for_an_account.feature (hunk 10 above)
Comment (HannaSwDn) [reply features/step_definitions/visitor_can_register_for_an_account.feature]: We have added tests for the sad path so that's why it is needed now

File: features/step_definitions/visitor_can_register_for_an_account.feature (hunk 11 above)
Comment (HannaSwDn) [reply features/step_definitions/visitor_can_register_for_an_account.feature]: Done

File: spec/models/user_spec.rb (hunk 8 above)
Comment (HannaSwDn) [reply spec/models/user_spec.rb]: Done

File: app/views/users/registrations/edit.html.erb (hunk 4 above)
Comment (HannaSwDn) [reply app/views/users/registrations/edit.html.erb]: Done

File: app/views/users/sessions/new.html.erb (hunk 3 above)
Comment (HannaSwDn) [reply app/views/users/sessions/new.html.erb]: Done

File: features/step_definitions/visitor_can_register_for_an_account.feature (hunk 12 above)
Comment (HannaSwDn) [inline features/step_definitions/visitor_can_register_for_an_account.feature]: Done

File: app/views/partials/_navbar.html.haml (hunk 13 above)
Comment (GergKllai1) [reply app/views/partials/_navbar.html.haml]: Like this
```
-if user_signed_in?
 =link_to "Logout", destroy_user_session_path, method: :delete
-else
 =link_to "Registration"
 =link_to "Login"

File: app/views/users/registrations/new.html.erb [hunk 15]
```diff
@@ -0,0 +1,31 @@
+%h2 Register
+
+= form_for(resource, as: resource_name, url: registration_path(resource_name)) do |f|
```
Comment (tochman) [inline app/views/users/registrations/new.html.erb:3]: You can make use of the new `form_with` helper

```
= form_with(model: resource, as: resource_name, url: registration_path(resource_name), remote: false) do |f|

```

File: features/step_definitions/basic_steps.rb [hunk 16]
```diff
@@ -15,6 +15,20 @@
...
 
+Given("the following user exists") do |table|
+    table.hashes.each do |user|
+        FactoryBot.create(:user, user)
+    end 
+end
+
+Given("I click {string}") do |string|
+    click_on string
+end
```
Comment (tochman) [inline features/step_definitions/basic_steps.rb:26]: Step duplication. See line 32 in this file.

File: spec/models/user_spec.rb [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+
+RSpec.describe User, type: :model do
+  describe 'DB table' do
+    it { is_expected.to have_db_column :email }
+    it { is_expected.to have_db_column :encrypted_password }
+  end
+
+  describe 'FactoryBot' do
+    it 'should have valid Factory' do
+      expect(FactoryBot.create(:user)).to be_valid
```
Comment (tochman) [inline spec/models/user_spec.rb]: Please include FactoryBot Matchers in your RSpec configuration (And Cucumber configuration) so that you can shorten this to
```
expect(create(:article)).to be_valid 
```

Comment (HannaSwDn) [reply spec/models/user_spec.rb]: Done

File: features/step_definitions/basic_steps.rb (hunk 16 above)
Comment (HannaSwDn) [reply features/step_definitions/basic_steps.rb:26]: Done

File: app/views/users/registrations/new.html.erb (hunk 15 above)
Comment (HannaSwDn) [reply app/views/users/registrations/new.html.erb:3]: We have been searching online and we do not see how it would benefit our code, could you please explain this new helper method to us?

File: Gemfile [hunk 18]
```diff
@@ -8,14 +8,15 @@ gem 'pg', '>= 0.18', '< 2.0'
...
-
+gem 'devise'
 gem 'coffee-rails', '~> 4.2'
 gem 'turbolinks', '~> 5'
 gem 'jbuilder', '~> 2.5'
 gem 'bootsnap', '>= 1.1.0', require: false
 gem 'webpacker', '~> 3.5'
 gem 'tailwindcss', '~> 0.2.0'
-gem "haml-rails", "~> 1.0"
+gem 'omniauth', '~> 1.6', '>= 1.6.1'
```
Comment (oliverochman) [inline Gemfile]: why do you add this gem?

File: features/visitor_can_register_for_an_account.feature [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+        And I fill in 'Password confirmation' field with 'password'
+        And I click 'Sign up'
+        Then I should see 'Welcome! You have signed up successfully.'
+
+    Scenario: When password is too short, I get an error message
+        Given I fill in 'Email' field with 'bill@email.com'
+        And I fill in 'Password' field with 'hello'
+        And I fill in 'Password confirmation' field with 'hello'
+        And I click 'Sign up'
+        And I should see 'Password is too short (minimum is 6 characters)'
```
Comment (oliverochman) [inline features/visitor_can_register_for_an_account.feature]: "Then I should see"
Instead of "Given I should see"

File: features/visitor_can_register_for_an_account.feature [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+
+    Scenario: When password confirmation doesn't match password, I get an error message
+        Given I fill in 'Email' field with 'bill@email.com'
+        And I fill in 'Password' field with 'password'
+        And I fill in 'Password confirmation' field with 'pasword'
+        And I click 'Sign up'
+        And I should see "Password confirmation doesn't match Password"
```
Comment (oliverochman) [inline features/visitor_can_register_for_an_account.feature]: "Then I should see"
Instead of "Given I should see"

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: Gemfile (hunk 18 above)
Comment (HannaSwDn) [reply Gemfile]: We are planning on adding login with LinkedIn later that's why we added it, but since we do not need it right now we will remove it.

File: features/visitor_can_register_for_an_account.feature (hunk 19 above)
Comment (HannaSwDn) [reply features/visitor_can_register_for_an_account.feature]: Done

File: features/visitor_can_register_for_an_account.feature (hunk 20 above)
Comment (HannaSwDn) [reply features/visitor_can_register_for_an_account.feature]: Done

**Review by oliverochman — APPROVED** [review]
